File: apps/server/src/storage/connections.ts
// connections.json — the saved connection list, written 0600 because it may hold passwords.

import { randomUUID } from "node:crypto";
import { type ConnectionConfig } from "@perch/protocol";
import { CONNECTIONS_FILE } from "./paths.js";
import { JsonStore } from "./json-file.js";

/** 0600: this is the one file perch writes that can hold a password. */
const store = new JsonStore<ConnectionConfig[]>(CONNECTIONS_FILE, () => [], 0o600);

/**
 * The connection without its password — the shape both the API's ConnectionSummary and the CLI's
 * `--json` output are built from. Secrets never leave this module by accident.
 */
export function redactConnection(config: ConnectionConfig): Omit<ConnectionConfig, "password"> {
  const rest: Partial<ConnectionConfig> = { ...config };
  delete rest.password;
  return rest as Omit<ConnectionConfig, "password">;
}

export const listConnections = (): Promise<ConnectionConfig[]> => store.read();

export const saveConnections = (list: ConnectionConfig[]): Promise<void> => store.write(list);

export async function getConnection(idOrName: string): Promise<ConnectionConfig | undefined> {
  const list = await listConnections();
  return list.find((c) => c.id === idOrName) ?? list.find((c) => c.name === idOrName);
}

export async function upsertConnection(
  input: Omit<ConnectionConfig, "id" | "createdAt"> & { id?: string },
): Promise<ConnectionConfig> {
  const list = await listConnections();
  const existing = input.id ? list.find((c) => c.id === input.id) : undefined;
  const record: ConnectionConfig = {
    ...(existing ?? { id: randomUUID(), createdAt: new Date().toISOString() }),
    ...input,
    id: existing?.id ?? input.id ?? randomUUID(),
  };
  const next = existing ? list.map((c) => (c.id === record.id ? record : c)) : [...list, record];
  await saveConnections(next);
  return record;
}

export async function removeConnection(idOrName: string): Promise<boolean> {
  const list = await listConnections();
  const next = list.filter((c) => c.id !== idOrName && c.name !== idOrName);
  if (next.length === list.length) return false;
  await saveConnections(next);
  return true;
}

File: apps/server/storage/connections.go
package storage

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"

	"perch/protocol"
)

// 0600: the one file perch writes that can hold a password.
var connectionsStore = NewStore(ConnectionsFile, 0o600, func() []protocol.ConnectionConfig {
	return []protocol.ConnectionConfig{}
})

func ListConnections() ([]protocol.ConnectionConfig, error) {
	return connectionsStore.Read()
}

func SaveConnections(list []protocol.ConnectionConfig) error {
	return connectionsStore.Write(list)
}

// GetConnection accepts either the id or the name, id first.
func GetConnection(idOrName string) (protocol.ConnectionConfig, bool, error) {
	list, err := ListConnections()
	if err != nil {
		return protocol.ConnectionConfig{}, false, err
	}
	for _, c := range list {
		if c.ID == idOrName {
			return c, true, nil
		}
	}
	for _, c := range list {
		if c.Name == idOrName {
			return c, true, nil
		}
	}
	return protocol.ConnectionConfig{}, false, nil
}

func UpsertConnection(record protocol.ConnectionConfig) (protocol.ConnectionConfig, error) {
	list, err := ListConnections()
	if err != nil {
		return record, err
	}
	for i, c := range list {
		if record.ID != "" && c.ID == record.ID {
			record.CreatedAt = c.CreatedAt
			list[i] = record
			return record, SaveConnections(list)
		}
	}
	if record.ID == "" {
		record.ID = NewID()
	}
	if record.CreatedAt == "" {
		record.CreatedAt = protocol.Now()
	}
	return record, SaveConnections(append(list, record))
}

func RemoveConnection(idOrName string) (bool, error) {
	list, err := ListConnections()
	if err != nil {
		return false, err
	}
	next := make([]protocol.ConnectionConfig, 0, len(list))
	for _, c := range list {
		if c.ID == idOrName || c.Name == idOrName {
			continue
		}
		next = append(next, c)
	}
	if len(next) == len(list) {
		return false, nil
	}
	return true, SaveConnections(next)
}

// NewID is a v4 UUID, matching the ids the TypeScript server wrote with randomUUID().
func NewID() string {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		return hex.EncodeToString(b[:])
	}
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}
